x/proof/keeper/msg_server_submit_proof_test.go

```diff
@@ -0,0 +1,1336 @@
+				unclaimedSessionTree := newFilledSessionTree(
+					ctx, t,
+					"wrong_supplier",
+					wrongSupplierAddr,
+					unclaimedSessionHeader,
+					unclaimedSessionHeader,
+					unclaimedSessionHeader,
+					keyRing,
+					ringClient,
+					5,
+				)
+
+				// Discard session tree Merkle root because no claim is being created.
+				// Session tree must be closed (flushed) to compute closest Merkle Proof.
+				_, err = unclaimedSessionTree.Flush()
+				require.NoError(t, err)
+
+				// Construct new proof message using the supplier & session header
+				// from the session which is *not* expected to be claimed.
+				return newTestProofMsg(t,
+					wrongSupplierAddr,
+					unclaimedSessionHeader,
+					unclaimedSessionTree,
+					expectedClosestMerkleProofPath,
+				)
+			},
+			expectedErr: status.Error(
+				codes.FailedPrecondition,
+				types.ErrProofClaimNotFound.Wrapf(
+					"no claim found for session ID %q and supplier %q",
+					unclaimedSessionHeader.GetSessionId(),
+					wrongSupplierAddr,
+				).Error(),
+			),
+		},
+		{
+			desc: "claim and proof session start heights must match",
+			newProofMsg: func(t *testing.T) *types.MsgSubmitProof {
+				// Advance the block height such that no hydration errors can occur when
+				// getting a session with start height less than the current block height.
+				sdkCtx := cosmostypes.UnwrapSDKContext(ctx)
+				ctx = sdkCtx.WithBlockHeight(3)
+
+				// Shift the session start height ...
+
+				t.Cleanup(func() {
+					// Restore the block height of the context to zero.
+					sdkCtx := cosmostypes.UnwrapSDKContext(ctx)
+					ctx = sdkCtx.WithBlockHeight(0)
+				})
+
+				// Construct new proof message.
+				return newTestProofMsg(t,
+					supplierAddr,
+					&wrongSessionStartHeightHeader,
+					validSessionTree,
+					expectedClosestMerkleProofPath,
+				)
+			},
+			expectedErr: status.Error(
+				codes.FailedPrecondition,
+				types.ErrProofInvalidRelay.Wrapf(
+					"session headers session start heights mismatch; expected: %d, got: %d",
+					2,
+					1,
+				).Error(),
+			),
+		},
+		{
+			desc: "claim and proof session end heights must match",
+			newProofMsg: func(t *testing.T) *types.MsgSubmitProof {
+				// Advance the block height such that no hydration errors can occur when
+				// getting a session with start height less than the current block height.
+				setBlockHeight(&ctx, 3)
+				// Reset the block height to zero after this test case.
+				t.Cleanup(resetBlockHeightFn(&ctx))
+
+				// Construct new proof message.
+				return newTestProofMsg(t,
+					supplierAddr,
+					&wrongSessionEndHeightHeader,
+					validSessionTree,
+					expectedClosestMerkleProofPath,
+				)
+			},
+			expectedErr: status.Error(
+				codes.FailedPrecondition,
+				types.ErrProofInvalidRelay.Wrapf(
+					"session headers session end heights mismatch; expected: %d, got: %d",
+					3,
+					4,
+				).Error(),
+			),
+		},
+		{
+			desc: "merkle proof must validate claimed merkle root",
```

Comment:
Can you update this? Seems like "Valid proof cannot validate claim with invalid root".